primitives/message-lane/src/lib.rs

```diff
@@ -23,53 +23,78 @@
 #![allow(clippy::unnecessary_mut_passed)]
 
 use codec::{Decode, Encode};
+use frame_support::{weights::Weight, RuntimeDebug};
 use sp_api::decl_runtime_apis;
+use sp_std::prelude::*;
+
+pub mod source_chain;
+pub mod target_chain;
 
 /// Lane identifier.
 pub type LaneId = [u8; 4];
 
 /// Message nonce. Valid messages will never have 0 nonce.
 pub type MessageNonce = u64;
 
+/// Message id as a tuple.
+pub type MessageId = (LaneId, MessageNonce);
+
 /// Message key (unique message identifier) as it is stored in the storage.
-#[derive(Encode, Decode, Clone)]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
 pub struct MessageKey {
 	/// ID of the message lane.
 	pub lane_id: LaneId,
 	/// Message nonce.
 	pub nonce: MessageNonce,
 }
 
+/// Message data as it is stored in the storage.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
+pub struct MessageData<Payload, Fee> {
+	/// Message payload.
+	pub payload: Payload,
+	/// Message delivery and dispatch fee, paid by the submitter.
+	pub fee: Fee,
+}
+
 /// Message as it is stored in the storage.
-#[derive(Encode, Decode, Clone)]
-pub struct Message<Payload> {
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
+pub struct Message<Payload, Fee> {
 	/// Message key.
 	pub key: MessageKey,
-	/// Message payload.
-	pub payload: Payload,
+	/// Message data.
+	pub data: MessageData<Payload, Fee>,
 }
 
 /// Message processing result.
-pub enum MessageResult<Payload> {
+#[derive(RuntimeDebug, PartialEq)]
+pub enum MessageResult<Payload, Fee> {
 	/// Message has been processed and should not be queued.
-	Processed,
+	Processed(Weight),
 	/// Message has NOT been processed and should be queued for processing later.
-	NotProcessed(Message<Payload>),
+	NotProcessed(Message<Payload, Fee>),
 }
 
-/// Called when inbound message is received.
-pub trait OnMessageReceived<Payload> {
-	/// Called when inbound message is received.
-	///
-	/// It is up to the implementers of this trait to determine whether the message
-	/// is invalid (i.e. improperly encoded, has too large weight, ...) or not. And,
-	/// if message is invalid, then it should be dropped immediately (by returning
-	/// `MessageResult::Processed`), or it'll block the lane forever.
-	fn on_message_received(&mut self, message: Message<Payload>) -> MessageResult<Payload>;
+impl<Payload, Fee> MessageResult<Payload, Fee> {
+	/// Return number of processed messages (0 or 1).
+	pub fn messages_processed(&self) -> MessageNonce {
+		match *self {
+			MessageResult::Processed(_) => 1,
+			MessageResult::NotProcessed(_) => 0,
+		}
+	}
```

Comment:
Is this ever used?

Also, I don't quite get the logic here. Why do we only ever process one message?

Reply:
> Is this ever used?

Not anymore - I'm removing it.

> Also, I don't quite get the logic here. Why do we only ever process one message?

Because `MessageResult` is result of processing single message.